packages/material-ui/src/ButtonBase/ButtonBase.js

```diff
@@ -89,7 +89,18 @@ class ButtonBase extends React.Component {
   });
 
   componentDidMount() {
-    prepareFocusVisible(this.getButtonNode().ownerDocument);
+    const button = this.getButtonNode();
+    if (button == null) {
```

Comment:
It looks like we only use the ref in the call to `prepareFocusVisible`, `handleKeyDown`, and the imperative `focusVisible`. If we removed the warning in `componentDidMount` and skipped `prepareFocusVisible` when no ref, we could still do the meaningful hard error in `handleKeyDown` and the imperative `focusVisible`. Presumably in the cases that actually need the warning they would still have the ref warning, they just won't get as clear of a message about what they should do.

This potentially leaves a hole where if the ref is missing for some other reason (e.g. a component that renders initially as null and then renders to the dom later), they won't receive a warning but the focus visible logic might not work correctly (since the listeners wouldn't have been set up).

Reply:
> This potentially leaves a hole where if the ref is missing for some other reason (e.g. a component that renders initially as null and then renders to the dom later)

Yeah handling instances from a ref only once is always a bit brittle. In the case of a button which is already very close to the DOM I don't think that's an issue though. I can't imagine valid use cases where the owner document of a button changes during its lifetime.

I still prefer the hard-error here since focus-visible is an important a11y feature. I'm not in favor of silencing a11y issues. We should actively notify app authors if their code leads to a11y issues.

Reply:
> I still prefer the hard-error here since focus-visible is an important a11y feature. I'm not in favor of silencing a11y issues.

The hard error seems too strict to have in `componentDidMount` since it will always fail with react-test-renderer. We could do a warning in `componentDidMount` (skipping `prepareFocusVisible`) that at least gives the clearer indication of the problem than React's ref warning. Only issue is if the user has their tests setup to fail on warnings, but that is at least a problem that can be worked around.

Reply:
> The hard error seems too strict to have in componentDidMount since it will always fail with react-test-renderer.

A warning might be missed and doesn't seem appropriate since we definitely know that the component will not work as expected. Supporting react-test-renderer should have a very low priority. It should definitely not trump a11y issues.

I'm having a hard time understanding the concerns. On one hand it's apparently acceptable to ship code that does not work until someone reports it. On the other hand we purposely silence actual errors because some testing environments fail if they're not set up properly. This seems like we're writing code to make devs comfortable (in a bad way) instead of writing code for the end-user.

Reply:
> Supporting react-test-renderer should have a very low priority.

I don't know what percentage of Material-UI users use react-test-renderer, but it seems like a very big barrier to migration to v4 if someone is heavily using react-test-renderer. It would basically mean saying in our migration guide that if you are using react-test-renderer, you need to redo all your tests to use something else. Given that react-test-renderer has [1,779,519 weekly downloads on npm](https://www.npmjs.com/package/react-test-renderer), I think it would be unwise to create a large barrier for all of those people to upgrade to v4.

Reply:
Sure but as I said this is not an issue specific to ButtonBase. We have a hard dependency on the DOM so if you test our components without mocking a DOM then errors should be expected. `MenuList` has the same issue. Probably others as well. 

It feels like there should be some library that handles mocking a DOM for `react-test-renderer`.

Reply:
Will the react-test-renderer issue still exist once this is converted to a function component? I think react-test-renderer only executes `useLayoutEffect`. If `prepareFocusVisible` is called in `useEffect` we could do a hard error without impacting react-test-renderer. Best of both worlds?

Reply:
> We have a hard dependency on the DOM so if you test our components without mocking a DOM then errors should be expected.

Though `ButtonBase` is used so pervasively that it will impact a lot more people (and a lot higher percentage of their tests) than any other components with similar issues.

Reply:
`react-test-renderer` has its own `act` which will cause the same issues. I'm still very uncomfortable with making simple<sup>1</sup> devs first-class citizen over users. Hiding this error means hiding an actual a11y issue that is more likely to occur with the switch to refs. 

I think we should put more effort into teaching people how to properly setup their tests instead of treating this behavior as a feature. Especially because it's an a11y concern and not just some animated background gradient that changes according to time of day and season or similar fancy stuff. 

<sup>1</sup> I don't mean this in a derogatory way. I like my test setups simple as well. But I'd rather have someone throw an error in my face if I make a mistake instead of asking me if I know what I'm doing in a 5 page letter.

I'm happy to work on a `createNodeMock` function that works for all `@material-ui/core` components. But since this is an issue not specific to the ButtonBase we can tackle this in a different PR.

Reply:
The developer's users happiness is our ultimate function of success. However, we have this incredible leverage: save a developer one hour, you will save x10 more to this developer's users. It's a difficult equilibrium. Make the tests hard to write for a developer, he won't write them.

I believe that https://github.com/mui-org/material-ui/pull/15627#issuecomment-490654600 can unify the concerns. We get the error as well as the simple snapshot test DX. Now, there are other dimensions to take into account
- Snapshot tests are already close to useless today with v3 outside of the box. Our class names are nondeterministic. People need to configure them. I don't know how bad the experience is as I never write any snapshot tests for UI components.
- This leads to my second point, should we encourage snapshot tests with UI elements. When I ask people retrospective experience about them, it's often: it's a waste of time, didn't catch bugs, we accept the changes without looking, we have too many of them.

Reply:
I outlined already why point 2 will immediately negate point 1. Effects can be run in react-test-renderer so there's nothing gained here. I will look into a createNodeMock but outside of that there's currently no viable solution.

Reply:
by default?

Reply:
https://reactjs.org/docs/test-utils.html#act

Reply:
https://jestjs.io/docs/en/snapshot-testing#snapshot-testing-with-jest

Reply:
It might not be an issue with the non-snapshot API because the ref value is correctly computed?

Reply:
Just verified that `react-test-renderer` will also crash with MenuList#autoFocus. I don't want to spend anymore time on finding instances where this crashes. I'm a bit confused why the dependency on the DOM is an issue all of the sudden. Will look into a createNodeMock since this seems important.

Reply:
The effects are not run by default: https://codesandbox.io/s/ypoj2y3px1?module=%2Fsrc%2Fmain.test.js&previewwindow=tests

react-test-renderer doesn't even [document](https://reactjs.org/docs/test-renderer.html) the `act` method except as a side note in [ReactTestUtils](https://reactjs.org/docs/test-utils.html#act).

I certainly don't think there is any need to support `act` working in react-test-renderer. I just think a lot of people have tests that do the initial rendering of their components as a sniff test to make sure they haven't broken something obvious. I agree with the sentiment that snapshot testing is not particularly valuable (and I think the community is in the process of moving towards move valuable testing approaches with things like react-testing-library), but it seems pretty harmless to do this in a way that will allow the `react-test-renderer` `create` function to continue working with `ButtonBase`.

Reply:
> MenuList has the same issue. Probably others as well.

@eps1lon I'm not advocating that we try to have the whole library support react-test-renderer. `Menu` would have already had this issue in v3, so we aren't likely impacting anyone new there. I am mainly just concerned about `ButtonBase` being a **new** place where this is an issue. Just about every page in most apps will have buttons and other things that leverage `ButtonBase`. I would expect that it is **the** most-used component.

Reply:
@ryancogswell I think that it's fine to give core contributors a free pass when they feel strongly about something. The recovery cost is cheap with this change. If we really need to change the approach in the future, we can. The advantage of being strict is that we will leanr more about people use cases. Developers can always open a new issue about the problem, as well as upvote it.

Reply:
react-test-renderer runs layout effects though which means e.g. a MenuList would crash because focus isn't mocked. I created a [small repo to illustrate the issues](https://github.com/eps1lon/material-ui-testing) and I can only come to the conclusion that react-test-renderer is not viable for our components. I couldn't get it to work for Modal.

Reply:
Aside: tests already crash in codesandbox when importing our components. We should also investigate that.

Reply:
I'm fine with that. I just think this would have been something that would have been nice for users to have longer warning about from a migration standpoint. I agree that it will be useful to know how many people really have a problem with this and how painful it will or won't be for them to address.